lib/spack/spack/solver/asp.py

```diff
@@ -4625,11 +4665,12 @@ def solve_with_stats(
         setup = SpackSolverSetup(tests=tests)
         output = OutputConfiguration(timers=timers, stats=stats, out=out, setup_only=setup_only)
 
-        CONC_CACHE.flush_manifest()
-        CONC_CACHE.cleanup()
-        return self.driver.solve(
+        result = self.driver.solve(
             setup, specs, reuse=reusable_specs, output=output, allow_deprecated=allow_deprecated
         )
+        CONC_CACHE.flush_manifest()
```

Comment:
Shouldn't `flush` be done on every write? Then you don't need to manually sync and deal with exceptions that are raised between the writing the concretizer output to cache and updating the stats.

Reply:
Follow up comments from @haampie suggest the use of a `ccache` style model where we lock a "bucket" (or cache entry subdir) and track the stats at that scope, significantly limiting the occurrence of concurrent writes causing a bottleneck. The same overarching cache statistics we currently track could be computed from each bucket via a method as part of the conc cache API. 

I think this is the best path forward, currently implementing and testing.